File: python/wrpc/manager/provider.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

'''
@author: shuai.chen
Created on 2017年3月4日
'''
from __future__ import absolute_import

from abc import ABCMeta, abstractmethod
import logging
import threading

from kazoo.exceptions import SessionExpiredError
from kazoo.recipe.watchers import ChildrenWatch

from wrpc.common import constant, util, WrpcException
from wrpc.common.node import ServerNode
from .load_balance import RoundRobinLoad

logger = logging.getLogger(__name__)

class Provider(object):
    
    __metaclass__ = ABCMeta
    
    @abstractmethod
    def select(self):
        raise NotImplementedError

    @abstractmethod
    def get_services(self):
        raise NotImplementedError
    
    def listen(self):
        pass
    
    def set_client_pool(self, client_pool):
        pass
    
    def close(self):
        pass   

class AutoProvider(Provider):
    
    __lock = threading.RLock()
    
    def __init__(self, zk_client, global_service_name, version=constant.VERSION_DEFAULT,
                 services = [], load_balance = RoundRobinLoad):
        """
        auto server provider listened by ChildrenWatch
        @param zkclient: zookeeper connection client
        @param global_service_name: global service name
        @param version: server version default is 1.0.0
        @param services: service ifaces class
        @param load_balance: load balance class, RoundRobinLoad or RandomLoad, 
                             default is RoundRobinLoad
        """
        try:
            self.__zk_client = zk_client 
        except SessionExpiredError:
            logger.warn("Zookeeper Client session timed out.")    
            
        self.__global_service_name = global_service_name
        self.__version = version   
        self.__services = services
        self.__load_balance = load_balance()
        
        self.__all_nodes = {}
        self.__live_nodes = set()  
        self.client_pool = None
        
    def __get_parent_path(self):
        zsd = constant.ZK_SEPARATOR_DEFAULT
        if self.__global_service_name != "":
            return "{0}{1}{2}{3}".format(zsd, self.__global_service_name, zsd, self.__version)
        return "{0}{1}".format(zsd, self.__version)
    
    def listen(self):
        if self.__zk_client:
            ChildrenWatch(self.__zk_client, self.__get_parent_path(), func=self.__watch_server_node)
        
    def __watch_server_node(self, nodes):
        with self.__lock:
            self.__live_nodes.clear()
            
            if len(nodes) <= 0:
                logger.warn("server not found!") 
            else:                
                for node in nodes:
                    server_node = ServerNode(node)
                    self.__all_nodes[node] = server_node
                    self.__live_nodes.add(server_node)  
                
            self.__load_balance.set_nodes(self.__live_nodes)    
                
            if self.client_pool is not None:
                self.client_pool.clear_pool()    
            
        logger.info("Child node changed.")        
    
    def select(self):
        return self.__load_balance.get_node()
    
    def get_services(self):
        return self.__services
    
    def set_client_pool(self, client_pool):
        self.client_pool = client_pool
        
    def close(self):
        if self.client_pool:
            self.client_pool.clear_pool()
        if self.__zk_client:
            self.__zk_client.close()

class FixedProvider(Provider):
    
    def __init__(self, server_address, services = [], load_balance = RoundRobinLoad):
        """
        fixed server provider
        @param server_address: server adress  as string 'ip:port:weight' or 'ip:port'     
        @param service_ifaces: service ifaces class
        @param load_balance: load balance class, RoundRobinLoad or RandomLoad, 
                             default is RoundRobinLoad
        """   
        if not util.check_hosts(server_address):
            raise WrpcException("Illegal server_address for FixedProvider.")        
             
        self.__server_address = server_address
        self.__services = services
        self.__load_balance = load_balance()

        self.__live_nodes = set()
        self.__set_nodes()
        
    def __set_nodes(self):
        nodes = self.__server_address.split(",")
        for node in nodes:
            server_node = ServerNode(node)
            self.__live_nodes.add(server_node)  
        
        self.__load_balance.set_nodes(self.__live_nodes)       
    
    def select(self):
        return self.__load_balance.get_node()
    
    def get_services(self):
        return self.__services
            

File: go/wrpc/provider.go
package wrpc

/**
@author shuai.chen
@created 2020年1月8日
**/

import (
	"bytes"
	"sync"
	"strings"
	"log"
    "github.com/samuel/go-zookeeper/zk"
)

type Provider interface {
	GetServices() []Service
	SetClientPool(pool *ClientPool)
	SetNodes()
	Select() (*Node, error)
}

type BaseProvider struct {
	allNodes map[string]*Node
	liveNodes []*Node
	loadBalance LoadBalance
	services []Service
}

/**
AutoProvider
**/
type AutoProvider struct {
	BaseProvider
	
	zkc *ZkClient
    clientPool *ClientPool

	globalServiceName string
	version string
	
    mutex sync.Mutex
}

/*
default auto provider
@zkc：zookeeper客户端
@services：service 列表
@globalServiceName：全局service名称，和server保持一致
*/
func NewDefaultAutoProvider(zkc *ZkClient, services []Service, globalServiceName string) *AutoProvider {
	return NewAutoProvider(zkc, services, globalServiceName, RANDOM, "")
}

/*
auto provider
@zkc：zookeeper客户端
@services：service 列表
@globalServiceName：全局service名称，和server保持一致
@loadBalanceType：负载均衡类型
@version：版本号，若传空字串，则默认为1.0.0
*/
func NewAutoProvider(zkc *ZkClient, services []Service, globalServiceName string,
                     loadBalanceType int, version string) *AutoProvider {
    provider := AutoProvider{} 
    provider.zkc = zkc
    provider.services = services        
    provider.loadBalance = GetLoadBalance(loadBalanceType) 
    provider.globalServiceName = globalServiceName
    provider.version = version 	
    	if version == ""{
		provider.version = VERSION_DEFAULT
	} 
    	provider.allNodes = map[string]*Node{}                	
	return &provider
}

/*
获取 zk 注册地址父级目录
*/
func (ap *AutoProvider) GetParentPath() string{
	var b bytes.Buffer
    b.WriteString(ZK_SEPARATOR_DEFAULT)
    if ap.globalServiceName != ""{
	    b.WriteString(ap.globalServiceName)
	    b.WriteString(ZK_SEPARATOR_DEFAULT)
    }
    b.WriteString(ap.version)
    return b.String()
}

func (ap *AutoProvider) GetServices() []Service {
	return ap.services
}

func (ap *AutoProvider) SetNodes(){    
    path := ap.zkc.GetAbsolutePath(ap.GetParentPath())
    
    nodes, _, child_ch, err := ap.zkc.Conn.ChildrenW(path)
    if err != nil {
    	    log.Println("get zk path: ", path)
        log.Println("ERROR- ", err)
    }else{
	    ap.mutex.Lock()  //加锁
	    
	    ap.liveNodes = (ap.liveNodes)[0:0]
	    if len(nodes) <= 0{
	    	    log.Println("server not found!")
	    }
	    
	    for _, nodeStr := range nodes{
	        serverNode := ServerNode(nodeStr)
	        ap.allNodes[nodeStr] = serverNode
	        ap.liveNodes = append(ap.liveNodes, serverNode)
	    }
	    ap.loadBalance.SetNodes(ap.liveNodes) 
	    ap.clientPool.ClearPool() 
	    
	    	ap.mutex.Unlock() //解锁   
    }
    go ap.watch(child_ch)
}

// watch the child events
func (ap *AutoProvider) watch(child_ch <-chan zk.Event){
    for{
	    select {
	    	case e := <-child_ch:
	    	    if e.Type == zk.EventNodeChildrenChanged {
				ap.SetNodes() 			    	
				log.Println("Child node changed.") 
	    	    }else if e.Type == zk.EventNodeCreated{
				ap.SetNodes() 			    	
				log.Println("Child node created.") 
	    	    }else if e.Type == zk.EventNodeDeleted{
				ap.SetNodes() 			    	
				log.Println("Child node deleted.") 
	    	    }
	    }
    }	
}

func (ap *AutoProvider) SetClientPool(pool *ClientPool){
	ap.clientPool = pool
}

func (ap *AutoProvider) Select() (*Node, error){
	node, err := ap.loadBalance.GetNode()
	return node, err
}

/**
FixedProvider
**/
type FixedProvider struct {
	BaseProvider
	serverAddress string
}

/*
fixed provider
@services：service 列表
@serverAddress: server address  "ip:port:weight," or "ip:port,"
@loadBalanceType：负载均衡类型
*/
func NewFixedProvider(services []Service, serverAddress string, loadBalanceType int) *FixedProvider {
    loadBalance := GetLoadBalance(loadBalanceType)    
    bp := BaseProvider{services:services, loadBalance:loadBalance, allNodes:map[string]*Node{}}             	
	return &FixedProvider{BaseProvider:bp, serverAddress:serverAddress}
}

func (fp *FixedProvider) GetServices() []Service {
	return fp.services
}

func (fp *FixedProvider) SetNodes(){
	nodes := strings.Split(fp.serverAddress, ",")
    for _, nodeStr := range nodes{
        serverNode := ServerNode(nodeStr)
        fp.liveNodes = append(fp.liveNodes, serverNode)
    }
    fp.loadBalance.SetNodes(fp.liveNodes)    
}

func (fp *FixedProvider) SetClientPool(pool *ClientPool){

}

func (fp *FixedProvider) Select() (*Node, error){
	node, err := fp.loadBalance.GetNode()
	return node, err
}

